### PR TITLE
Tests for rejecting invites to unreachable servers

### PR DESCRIPTION
**Pull request by richvdh**

### CODE REVIEW

**Comment by richvdh** [comment]
retest this please

File: tests/50federation/00prepare.pl [hunk 1]
```diff
@@ -11,51 +11,57 @@
 
 my $DIR = dirname( __FILE__ );
 
-push our @EXPORT, qw( INBOUND_SERVER OUTBOUND_CLIENT );
+push our @EXPORT, qw( INBOUND_SERVER OUTBOUND_CLIENT create_federation_server );
+
+sub create_federation_server {
```
Comment (leonerd) [inline tests/50federation/00prepare.pl]: `{` wants to be on its own line for new toplevel functions.

File: tests/50federation/00prepare.pl [hunk 2]
```diff
@@ -11,51 +11,57 @@
...
+   my $server = SyTest::Federation::Server->new;
+   $loop->add( $server );
+
+   require IO::Async::SSL;
```
Comment (leonerd) [inline tests/50federation/00prepare.pl]: There's probably no need to hide this `require` within a function; just move it to toplevel at the start of the file

File: tests/50federation/35room-invite.pl [hunk 3]
```diff
@@ -142,50 +143,73 @@ sub invite_server
...
-
-               Future->done;
-            }),
-            matrix_leave_room( $user, $room_id )
-         )
-      })->then( sub {
-         matrix_sync( $user );
-      })->then( sub {
-         my ( $body ) = @_;
+foreach my $error_code (403, 500, -1) {
```
Comment (leonerd) [inline tests/50federation/35room-invite.pl]: Wants whitespace inside parens

    ( 403, 500, -1 )

File: tests/50federation/35room-invite.pl [hunk 4]
```diff
@@ -142,50 +143,73 @@ sub invite_server
...
+   # a temporary federation server which is shut down at the end of the test.
+   # we use a temporary server because otherwise the remote server ends up on the
+   # backoff list and subsequent tests fail.
+   my $temp_federation_server_fixture = fixture(
+      setup => sub {
+         create_federation_server()
+      },
+      teardown => sub {
+         my ($server) = @_;
+         $server -> close();
```
Comment (leonerd) [inline tests/50federation/35room-invite.pl]: No spaces around `->`

    $server->close();

File: tests/50federation/35room-invite.pl [hunk 5]
```diff
@@ -142,50 +143,73 @@ sub invite_server
...
+         $room->create_initial_events(
+            server  => $federation_server,
+            creator => $creator_id,
+         );
+
+         my $room_id = $room->room_id;
+
+         invite_server( $room, $creator_id, $user, $federation_server )
+         ->then( sub {
+            if ($error_code < 0) {
```
Comment (leonerd) [inline tests/50federation/35room-invite.pl]: if( $error_code < 0 ) {

File: tests/50federation/35room-invite.pl [hunk 6]
```diff
@@ -142,50 +143,73 @@ sub invite_server
...
+               # now shut down the remote server, so that we get an 'unreachable'
+               # error on make_leave
+               $federation_server->close();
+
+               return matrix_leave_room( $user, $room_id );
+            } else {
```
Comment (leonerd) [inline tests/50federation/35room-invite.pl]: Uncuddle the `else`

    }
    else {

File: tests/50federation/35room-invite.pl [hunk 7]
```diff
@@ -142,50 +143,73 @@ sub invite_server
...
+   );
 
-         log_if_fail "Sync body", $body;
-         assert_json_object( $body->{rooms}{invite} );
-         keys %{ $body->{rooms}{invite} } and die "Expected empty dictionary";
-         Future->done(1);
-      });
-   };
+   test "Inbound federation can receive invite and reject when "
+         . ($error_code >= 0 ? "remote replies with a $error_code" : "is unreachable" ),
```
Comment (leonerd) [inline tests/50federation/35room-invite.pl]: Whitespace

    ( $error_code ...

**Comment by leonerd** [comment]
Semantically seems good; even commented to say why each test needs its own server :)

Just needs some small formatting fixes.

File: tests/50federation/00prepare.pl (hunk 2 above)
Comment (richvdh) [inline tests/50federation/00prepare.pl]: mmkay. I was just cargo-culting the original (https://github.com/matrix-org/sytest/pull/353/files#diff-ae1f9e915c208806589880fe9bd6e5f5L21). Moved anyway.

**Comment by leonerd** [comment]
LGTM

**Comment by richvdh** [comment]
ta

**Pull request merged**